app/services/oidc_jwt_token_validate.rb

```diff
@@ -0,0 +1,56 @@
+class OidcJwtTokenValidate
+  attr_reader :decoded_token
+  delegate :logger, to: :Rails
+
+  def initialize(jwt_token)
+    @jwt_token = jwt_token
+  end
+
+  def valid_jwt?
+    valid_signature? && type_bearer? && valid_aud?
+  rescue JWT::InvalidAudError
+    logger.error "OidcJwtToken SSO: Invalid token."
+    false
+  rescue JWT::ImmatureSignature
+    logger.error "OidcJwtToken SSO: JWT is for the future."
+    false
+  rescue JWT::ExpiredSignature
+    logger.error "OidcJwtToken SSO: Signature has expired."
+    false
+  rescue JWT::DecodeError
+    logger.error "OidcJwtToken SSO: Failed to decode JWT."
+    false
+  end
+
+  private
+
+  def generate_public_key
+    response = RestClient::Request.execute(
+      :url => Setting['oidc_jwks_url'],
+      :method => :get,
+      :verify_ssl => true
+    )
+    rsa_public_json = JSON.parse(response)['keys'].first
+    jwk = JSON::JWK.new(rsa_public_json)
+    jwk.to_key
+  rescue RestClient::Exception, SocketError, JSON::ParserError => e
+    logger.error e
+    nil
+  end
+
+  def valid_signature?
+    @decoded_token = JWT.decode(@jwt_token, generate_public_key, true, { algorithm: 'RS256' }).first
+    @decoded_token.present?
+  rescue JWT::DecodeError => e
+    logger.error e
+    false
+  end
+
+  def type_bearer?
+    @decoded_token['typ'] == 'Bearer'
+  end
+
+  def valid_aud?
```

Comment:
The JWT library can handle the verification of the audience, see https://github.com/jwt/ruby-jwt#audience-claim